packaging/macos/utility.m

```diff
@@ -0,0 +1,205 @@
+
+	// TODO: This snippet increasing the open file limit was copied from
+	// the monodevelop launcher stub. It may not be needed for OpenRA.
+	struct rlimit limit;
+	if (getrlimit(RLIMIT_NOFILE, &limit) == 0 && limit.rlim_cur < 1024)
+	{
+		limit.rlim_cur = limit.rlim_max < 1024 ? limit.rlim_max : 1024;
+		setrlimit(RLIMIT_NOFILE, &limit);
+	}
+
+	void *libmono = dlopen([hostPath UTF8String], RTLD_LAZY);
+	if (!libmono)
+	{
+		NSLog(@"Failed to load libmonosgen-2.0.dylib: %s\n", dlerror());
+		return 1;
+	}
+
+	mono_main _mono_main = (mono_main)dlsym(libmono, "mono_main");
+	if (!_mono_main)
+	{
+		NSLog(@"Could not load mono_main(): %s\n", dlerror());
+		return 1;
+	}
+
+	// Insert hostpath, dll and modId as arguments. Overwrite the first argument which was used to launch this application.
+	for (int i = argc; i > 0; i--)
+		argv[i + 1] = argv[i - 1];
+
+	argv[0] = (char*)[hostPath UTF8String];
+	argv[1] = (char*)[dllPath UTF8String];
+	argv[2] = modId;
+	argc += 2;
+
+	return _mono_main(argc, argv);
+}
+
+int launch_dotnet(int argc, char **argv, char *modId, bool isArmArchitecture)
+{
+	NSString *exePath = [[[NSBundle mainBundle] bundlePath] stringByAppendingPathComponent: @"Contents/MacOS/"];
+	NSString *dllPath;
+	NSString *hostPath;
+
+	if (isArmArchitecture)
+	{
+		hostPath = [exePath stringByAppendingPathComponent: @"arm64/libhostfxr.dylib"];;
+		dllPath = [exePath stringByAppendingPathComponent: @"arm64/OpenRA.Utility.dll"];
+	}
+	else
+	{
+		hostPath = [exePath stringByAppendingPathComponent: @"x86_64/libhostfxr.dylib"];;
+		dllPath = [exePath stringByAppendingPathComponent: @"x86_64/OpenRA.Utility.dll"];
+	}
+
+	void *lib = dlopen([hostPath UTF8String], RTLD_LAZY);
+	if (!lib)
+	{
+		NSLog(@"Failed to load %@: %s\n", hostPath, dlerror());
+		return 1;
+	}
+
+	hostfxr_initialize_for_dotnet_command_line_fn hostfxr_initialize_for_dotnet_command_line = (hostfxr_initialize_for_dotnet_command_line_fn)dlsym(lib, "hostfxr_initialize_for_dotnet_command_line");
+	if (!hostfxr_initialize_for_dotnet_command_line)
+	{
+		NSLog(@"Could not load hostfxr_initialize_for_dotnet_command_line(): %s\n", dlerror());
+		return 1;
+	}
+
+	hostfxr_run_app_fn hostfxr_run_app = (hostfxr_run_app_fn)dlsym(lib, "hostfxr_run_app");
+	if (!hostfxr_run_app)
+	{
+		NSLog(@"Could not load hostfxr_run_app(): %s\n", dlerror());
+		return 1;
+	}
+
+	hostfxr_close_fn hostfxr_close = (hostfxr_close_fn)dlsym(lib, "hostfxr_close");
+	if (!hostfxr_close)
+	{
+		NSLog(@"Could not load hostfxr_close(): %s\n", dlerror());
+		return 1;
+	}
+
+	struct hostfxr_initialize_parameters params;
+	params.size = sizeof(params);
+	params.host_path = (char*)[[exePath stringByAppendingPathComponent: @"Utility"] UTF8String];
+	params.dotnet_root = dirname((char*)[hostPath UTF8String]);
+
+	// Insert dll and modId as arguments. Overwrite the first argument which was used to launch this application.
+	for (int i = argc; i > 0; i--)
+		argv[i] = argv[i - 1];
```

Comment:
I just noticed that this is overflowing the `argv` buffer for the last element. Bad idea! Please go back to allocating a new array (but the `memmove` is still not needed).